### PR TITLE
Fix crash when converting to json the result of an (i)map

### PR DESCRIPTION
**Pull request by sereinity**
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
`lib/ansible/module_utils/basic.py`

##### ANSIBLE VERSION
```
2.2.0.0
```

##### SUMMARY
Results of `map`s command (patched by `six` to become `imaps`) can't be converted to json. The current behavior breaks some modules as `haproxy`.
This patch makes the results of `imaps` intelligible as a sequence type.

### CODE REVIEW

**Comment by gundalow** [comment]
@sereinity What's the steps to reproduce this, how does it break haproxy?

File: lib/ansible/module_utils/basic.py [hunk 1]
```diff
@@ -80,7 +80,7 @@
 import platform
 import errno
 import datetime
-from itertools import repeat, chain
+from itertools import repeat, chain, imap
```
Comment (mattclay) [inline lib/ansible/module_utils/basic.py]: There is no `imap` in `itertools` in python 3:

http://stackoverflow.com/questions/30271712/i-cant-find-imap-in-itertools-in-python

Comment (sereinity) [reply lib/ansible/module_utils/basic.py]: Yes, I have seen this. It will be fixed Monday ; and I will provide a full test scenario.

**Comment by sereinity** [comment]
@gundalow So, to reproduce the error: Install a haproxy up and running.
Then to test the module :
```
ansible -i my_inventory haproxy-servers -m haproxy -a 'host=bk-test state=enabled socket=/var/run/haproxy/admin.sock'
test-haproxy | FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Shared connection to 172.22.13.47 closed.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_SgsObv/ansible_module_haproxy.py\", line 350, in <module>\r\n    main()\r\n  File \"/tmp/ansible_SgsObv/ansible_module_haproxy.py\", line 345, in main\r\n    ansible_haproxy.act()\r\n  File \"/tmp/ansible_SgsObv/ansible_module_haproxy.py\", line 299, in act\r\n    state_before = self.get_state_for(self.backend, self.host)\r\n  File \"/tmp/ansible_SgsObv/ansible_module_haproxy.py\", line 245, in get_state_for\r\n    data = self.execute('show stat', 200, False).lstrip('# ')\r\n  File \"/tmp/ansible_SgsObv/ansible_module_haproxy.py\", line 181, in execute\r\n    self.client.connect(self.socket)\r\n  File \"/usr/lib/python2.7/socket.py\", line 224, in meth\r\n    return getattr(self._sock,name)(*args)\r\nsocket.error: [Errno 2] No such file or directory\r\n",
    "msg": "MODULE FAILURE"
```

**Comment by sereinity** [comment]
@gundalow To reproduce the error, install a haproxy up and running, then :
```
ansible -i my_inventory haproxy-servers -m haproxy -a 'host=bk-test state=enabled'
test-haproxy | FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Shared connection to 172.22.13.47 closed.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_yDXXyX/ansible_module_haproxy.py\", line 350, in <module>\r\n    main()\r\n  File \"/tmp/ansible_yDXXyX/ansible_module_haproxy.py\", line 345, in main\r\n    ansible_haproxy.act()\r\n  File \"/tmp/ansible_yDXXyX/ansible_module_haproxy.py
\", line 317, in act\r\n    self.module.exit_json(**self.command_results)\r\n  File \"/tmp/ansible_yDXXyX/ansible_modlib.zip/ansible/module_utils/basic.py\", line 1799, in exit_json\r\n  File \"/tmp/ansible_yDXXyX/ansible_modlib.zip/ansible/module_utils/basic.py\", line 385, in remove_values\r\nTypeError: isinstance()
 arg 2 must be a class, type, or tuple of classes and types\r\n",
    "msg": "MODULE FAILURE"
}
```

**Comment by sereinity** [comment]
The change on the haproxy module needed and sufficient, but I think it would be better for everyone to support (i)map during json conversion and everything.

**Comment by gundalow** [comment]
Rebase needed

**Comment by sereinity** [comment]
@gundalow rebase done

**Comment by mscherer** [comment]
So I did tried to resolve the conflict, but then added 1 empty line. The code look good however.

File: lib/ansible/modules/network/haproxy.py [hunk 2]
```diff
@@ -288,13 +288,14 @@ def get_state_for(self, pxname, svname):
...
         r = csv.DictReader(data.splitlines())
+
         state = tuple(
             map(
                 lambda d: {'status': d['status'], 'weight': d['weight']},
                 filter(lambda d: (pxname is None or d['pxname'] == pxname) and d['svname'] == svname, r)
             )
         )
-        return state or None
+        return list(state) or None
```
Comment (abadger) [inline lib/ansible/modules/network/haproxy.py]: Why is this needed?  It looks like state should already be a tuple which is a Sequence type?

File: lib/ansible/module_utils/basic.py [hunk 3]
```diff
@@ -82,6 +82,12 @@
 import datetime
 from itertools import repeat, chain
 
+from ansible.module_utils.pycompat24 import get_exception, literal_eval
+from ansible.module_utils.six import (PY2, PY3, b, binary_type, integer_types,
+        iteritems, text_type, string_types)
+from ansible.module_utils.six.moves import map, reduce
+from ansible.module_utils._text import to_native, to_bytes, to_text
+
```
Comment (abadger) [inline lib/ansible/module_utils/basic.py]: This should not be moved.  They're libraries from the current project so go below the imports for stdlib and third party libraries.

File: lib/ansible/module_utils/basic.py [hunk 4]
```diff
@@ -115,9 +121,9 @@
 # this matters, make sure to check for strings before checking for sequencetype
 try:
     from collections.abc import KeysView
-    SEQUENCETYPE = (Sequence, KeysView)
+    SEQUENCETYPE = (Sequence, KeysView, map)
```
Comment (abadger) [inline lib/ansible/module_utils/basic.py]: This is wrong as map is not a Sequence.  Pehaps we need to be checking for Iterables in some specific sections of code instead of Sequences?

**Comment by abadger** [comment]
The error quoted in https://github.com/ansible/ansible/pull/20644#issuecomment-276062018 also looks unrelated to the fixes proposed here.  The error is being raised because isinstance() is not being fed the right second argument.  The changes in the haproxy module affect the first argument.  The changes in basic.py affect the second argument but not in a way that should be able to fix the error.

**Comment by abadger** [comment]
Regarding the comment about list()/tuple() in haproxy.py:

Looks like tuple was introduced to fix this same error in this PR: https://github.com/ansible/ansible-modules-extras/pull/3350

If someone is interested in splitting an ITERABLETYPE from SEQUENCETYPE and updating code in basic.py to use the one that is appropriate, that can be build out of this PR.  Otherwise, this PR can be closed as the bug report prompting it has been addressed.

needs_revision

File: lib/ansible/module_utils/basic.py [hunk 5]
```diff
@@ -167,6 +167,8 @@
 from ansible.module_utils.six.moves import map, reduce
 from ansible.module_utils._text import to_native, to_bytes, to_text
 
+ITERABLETYPE = (map,)
+
```
Comment (abadger) [inline lib/ansible/module_utils/basic.py:171]: Note: I was thinking more along the lines of:
``` python

# Note: strings are Sequences and all sorts of things (including dictionaries and strings) are iterables.
# So if differentiating between these, it may make a big difference the order that the code checks it in.
# To differentiate those mentioned types, check in this order: [string_types, Sequences, Mappings, Iterables] 
try:
    from collections import iterable, Sequence, Mapping
except ImportError:
    # python2.5
    Sequence = (list, tuple)
    Mapping = (dict,)
    Iterable = (list, tuple, map)
``` 

Thinking about this even more, I think you can probably get rid of SEQUENCETYPE altogether as long as the places where it should accept a KeyView should be taking Iterable instead (KeyView is actually an iterable but not a Sequence.  I think the current code is abusing that.)

**Comment by sereinity** [comment]
It seems that another commit fixes haproxy for me, so I don't need this PR.

The change to allow Iterables to be jsonified no longer works due to not using `return_values` anymore.

It seems that this integration cost too much with no direct features (except returning iterables in modules).

To make it works we need to parse every field in returned values and 'realize' iterables.

@abadger Is it OK to you ? (Or anyone else)

**Comment by abadger** [comment]
That seems a bit costly without any solid reasons for implementing it.  Since haproxy now works I'd probably incline towards not implementing this in basic.py until/unless it is shown to be a real problem for other modules.

**Comment by sereinity** [comment]
agreed.
